external/shellclient/v1/infinitepipe.go

```diff
@@ -0,0 +1,112 @@
+package v1
+
+import (
+	"io"
+	"sync"
+)
+
+// InfinitePipeReader is a reading side of an InfinitePipe, similar to io.PipeReader.
+type InfinitePipeReader struct {
+	*InfinitePipeWriter
+}
+
+// InfinitePipeWriter is a writing side of an InfinitePipe, similar to io.PipeWriter.
+type InfinitePipeWriter struct {
+	c      sync.Cond
+	m      sync.Mutex
+	buffer []byte
+	closed bool
+	broken bool
+	tell   chan<- bool
+}
+
+// InfinitePipe is similar to ioext.AsyncPipe() except that writes will always
+// succeed. Data will be added to an internal buffer that can grow bigger than
+// the specified capacity.
+// Additionally, you may supply a channel tell that will be told whenever
+// the Infinite channel has been emptied, so that more bytes can be requested
+// to be written.
+//
+// This pipe kind is useful when implementing simple congestion control.
+func InfinitePipe(tell chan<- bool) (*InfinitePipeReader, *InfinitePipeWriter) {
```

Comment:
So this pipe never does `tell<- ...` anywhere....

So I suspect you have two pipes because:

A) For stdout/stderr (ie. the incoming streams), we need a to write the messages to a pipe that has an infinite internal buffer, and when the buffer is above a certain threshold you want to send "pause", when the buffer is empty you want to send "resume" with the respective `streamID == stdout or stderr`.

It seems to me that `DrainingPipe` is good for this, right? When you receive a message from `tell` the pipe is empty and we should send a "resume" message... If when writing to `DrainingPipe` you get a `ErrPipeFull` error then you should send a "pause" message.

B) For stdin (ie. the outgoing stream), we probably need a pipe we can block... Say we call it `r, w := BlockablePipe()`.
Then it should always be safe to call `r.Block()` to block further reads from the pipe, and `r.Unblock()` to unblock further reads...
In unblocked mode the pipe just acts like a normal pipe... In blocked mode writes are blocked.

So you would have one thread reading from `r` and writing messages to the websocket... and if in the receiving thread you get a message to "pause" stdin, then you call `r.Block()` if you get a message to "resume" stdin, then you call `r.Unblock()`.

I think you could do this by simply wrapping an existing `io.Pipe()` and then have a `sync.Cond/sync.Mutex` and a `bool` indicating if it's blocked or not...